### ISSUE TITLE
Wallet can sign an offer without requiring a roundtrip to the chain

### ISSUE THREAD

**Issue by rowgraus** [Agoric/Agoric#3908]
Acceptance Criteria:

- After clicking Swap on the AMM, offer arrives in the wallet within 3 seconds

**Comment by dckc** [comment]
I wonder if what @dtribble just describes works for the voting case: #3901

I also wonder how this compares to web3 APIs.

**Comment by rowgraus** [comment]
Acceptance Criteria:
* After clicking Swap on the AMM, offer arrives in the wallet within 3 seconds

**Comment by dckc** [comment]
status quo:

```mermaid
sequenceDiagram
    dApp->>bridge: getPublicFacet
    bridge->>dApp: publicFacet
    dApp->>publicFacet: getVaultInvitation
    publicFacet->>dApp: invitation
    dApp->>bridge: addOffer(invitation, ...)
    bridge->>user: ok?
    user->>bridge: ok.
    bridge->>zoe: offer(invitation, ...)
```

design sketch:

```mermaid

sequenceDiagram
    dApp->>bridge: addOffer({contractPub: "board2342", method: "makeVaultInvitation", give, want})
    bridge->>user: {give, want...} ok?
    user->>bridge: ok.
    bridge->>SmartWallet: addOffer({public: "board2342", method: "makeVaultInvitation"})
    SmartWallet->>zoe: getPublicFacet
    zoe->>SmartWallet: publicFacet
    SmartWallet->>publicFacet: getVaultInvitation
    SmartWallet->>zoe: offer(invitation, ...)
```

**Comment by dckc** [comment]
refinement, based on discussion with @michaelfig 

```mermaid

sequenceDiagram

participant user
participant RPC
participant walletUI
participant walletSt
participant walletB
participant dApp

    walletUI->>RPC: ?: published.wallet.agoric1..
    RPC-->>walletUI: stateN
    dApp->>+walletB: leader = makeLeader(wk-url)
      note left of dApp: creates an iframe to well-known url powerbox.agoric.app
    dApp->>dApp: f = makeFollower(leader)
    dApp->>dApp: c = leader.makeClient()
    dApp->>walletB: f.next()
    walletB->>walletSt: getDappState
    user->>walletUI: approve dApp
    walletSt-->>walletB: filter(stateN, dApp)
    walletB-->>dApp: filter(stateN, dApp)
    note left of dApp: stateN includes services.board, .zoe
    dApp->>walletB: instanceP <- E(E(client).getBoard()).getValue('asdf')
    dApp->>walletB: pfP <- E(E(client).getZoe()).getPublicFacet(instanceP)
    user->>dApp: send me tokens
    dApp->>walletB: invP <- E(pfP).makeInvitation()
    dApp->>walletB: E(c).deposit(invP)
    dApp->>walletB: offerIdP <- E(c).propose({want: 5tok})
    note right of walletB: offerId is edge name e.g.: 45
    walletB->>walletUI: proposal
    note right of walletUI: hits storage
    walletUI->>user: proposal
    user->>walletUI: Approve
    walletUI->>RPC: tx([trivialActions, acceptZoeOffer])
    RPC->>walletUI: stateN'
    walletUI->>walletSt: stateN'
    walletSt->>walletB: stateN'
    note right of walletB: stateN' includes offerstates
    walletB->>dApp: filter(stateN', dApp)
 
```


## Description of the Design

dapp will use `@agoric/casting` library to interact with the wallet front-end. (currently uses `@endo/captp`)

casting has
 - a single store for outbound messages (= current state of the stream)
 - a single store for inbound messages (= requests to change the state of the stream)
   - can only influence state of stream; not much response; only a broken promise on connection failure; no tx failure notice
   - promises can be pipelines, but you can't get values from them other than by following the stream

transaction manager (_TODO: replace with relevant DB terminology_)
 - strict
 - speculative
 - optimistic

**Comment by dckc** [comment]
fixed by #5946
